File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.java
//package org.firstinspires.ftc.teamcode;
//
//import com.qualcomm.robotcore.eventloop.opmode.Autonomous;
//import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
//import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
//import com.qualcomm.robotcore.hardware.DcMotor;
//import com.qualcomm.robotcore.hardware.HardwareMap;
//import com.qualcomm.hardware.rev.RevBlinkinLedDriver;
//import com.vuforia.Vuforia;
//
//import org.firstinspires.ftc.robotcore.external.ClassFactory;
//import org.firstinspires.ftc.robotcore.external.navigation.VuforiaLocalizer;
//import org.firstinspires.ftc.robotcore.external.tfod.Recognition;
//import org.firstinspires.ftc.robotcore.external.tfod.TFObjectDetector;
//
//import java.util.List;
//
//class MecanumDrive {
//    private DcMotor frontLeft;
//    private DcMotor frontRight;
//    private DcMotor rearLeft;
//    private DcMotor rearRight;
//    private DcMotor lift;
//    private RevBlinkinLedDriver blinkinLedDriver;
//
//
//    MecanumDrive(HardwareMap hardwareMap) {
//        frontLeft = hardwareMap.get(DcMotor.class, "front_left");
//        frontRight = hardwareMap.get(DcMotor.class, "front_right");
//        rearLeft = hardwareMap.get(DcMotor.class, "rear_left");
//        rearRight = hardwareMap.get(DcMotor.class, "rear_right");
//
//        frontRight.setDirection(DcMotor.Direction.REVERSE);
//        rearRight.setDirection(DcMotor.Direction.REVERSE);
//
//        lift = hardwareMap.get(DcMotor.class, "lift");
//
//        // blinkinLedDriver = hardwareMap.get(RevBlinkinLedDriver.class, "blinkin");
//        // blinkinLedDriver.setPattern(RevBlinkinLedDriver.BlinkinPattern.RAINBOW_RAINBOW_PALETTE);
//    }
//
//    void drive(double x, double y, double z) {
//        frontLeft.setPower(y + z + x);
//        frontRight.setPower(y - z - x);
//        rearLeft.setPower(y + z - x);
//        rearRight.setPower(y - z + x);
//    }
//
//    void moveLift(double power) {
//        lift.setPower(power);
//    }
//}
//
//@TeleOp(name="Mecanum TeleOp", group="TeleOp")
//class MecanumTeleOp extends LinearOpMode {
//    @Override
//    public void runOpMode() {
//        MecanumDrive driver = new MecanumDrive(hardwareMap);
//
//        waitForStart();
//
//        while (opModeIsActive()) {
//            double x =-gamepad1.left_stick_x;
//            double y = gamepad1.left_stick_y;
//            double z = -gamepad1.right_stick_x;
//            driver.drive(x, y, z);
//            driver.moveLift(gamepad2.left_stick_y);
//        }
//    }
//}
//
//@Autonomous(name="Mecanum Autonomous", group="Autonomous")
//class MecanumAutonomus extends LinearOpMode {
//    private static final String TFOD_MODEL_ASSET = "RoverRuckus.tflite";
//    private static final String LABEL_GOLD_MINERAL = "Gold Mineral";
//    private static final String LABEL_SILVER_MINERAL = "Silver Mineral";
//    private static final String VUFORIA_KEY = "AWalSLT/////AAABme4BfTt4y0Wnrn7kzOornkwjbHAMMZnWYYe0WOYGLbmQ914wZ8gozbWoSgsLHUQ+asiEx9VUOuDMkr4LsB5hB3iyn/JZM7NB1fG15dZSIXVDXevQ2iy+6zaVS8q0JZ2dpvugMWHEpsrmadtlA13znd7nTwhvVoc+2gGc2hNb5k6G4qe4l3jusYi3KGXTkzLNvjjYuQGP8xsBkGkXY3oHt6LNwdBy9EAcxRYmKEQUoGOOSNAL5vqB2ZHN2pKrlxXJs0BMbyhm1U4PTPYriGCb2rBmgiBvcJmfz9T/vVLCpL3T9mV1GByU38J1aMzj6QrQKWD6/ImZYMchpb2yG5Xq0vb1WyFGR9IKp6eaAJ0XQL+q";
//
//    @Override
//    public void runOpMode() {
//        // MecanumDrive driver = new MecanumDrive(hardwareMap);
//
//        VuforiaLocalizer.Parameters parameters = new VuforiaLocalizer.Parameters();
//        parameters.vuforiaLicenseKey = VUFORIA_KEY;
//        parameters.cameraDirection = VuforiaLocalizer.CameraDirection.BACK;
//        VuforiaLocalizer vuforia = ClassFactory.getInstance().createVuforia(parameters);
//
//        int tfodMonitorViewId = hardwareMap.appContext.getResources().getIdentifier(
//                "tfodMonitorViewId", "id", hardwareMap.appContext.getPackageName());
//        TFObjectDetector.Parameters tfodParameters = new TFObjectDetector.Parameters(tfodMonitorViewId);
//        TFObjectDetector tfod = ClassFactory.getInstance().createTFObjectDetector(tfodParameters, vuforia);
//        tfod.loadModelFromAsset(TFOD_MODEL_ASSET, LABEL_GOLD_MINERAL, LABEL_SILVER_MINERAL);
//
//        waitForStart();
//
//        while (opModeIsActive()) {
//            // driver.drive(0, 0, 0);
//
//            if (tfod != null) {
//                // getUpdatedRecognitions() will return null if no new information is available since
//                // the last time that call was made.
//                List<Recognition> updatedRecognitions = tfod.getUpdatedRecognitions();
//                if (updatedRecognitions != null) {
//                    telemetry.addData("# Object Detected", updatedRecognitions.size());
//                    if (updatedRecognitions.size() == 3) {
//                        int goldMineralX = -1;
//                        int silverMineral1X = -1;
//                        int silverMineral2X = -1;
//                        for (Recognition recognition : updatedRecognitions) {
//                            if (recognition.getLabel().equals(LABEL_GOLD_MINERAL)) {
//                                goldMineralX = (int) recognition.getLeft();
//                            } else if (silverMineral1X == -1) {
//                                silverMineral1X = (int) recognition.getLeft();
//                            } else {
//                                silverMineral2X = (int) recognition.getLeft();
//                            }
//                        }
//                        if (goldMineralX != -1 && silverMineral1X != -1 && silverMineral2X != -1) {
//                            if (goldMineralX < silverMineral1X && goldMineralX < silverMineral2X) {
//                                telemetry.addData("Gold Mineral Position", "Left");
//                            } else if (goldMineralX > silverMineral1X && goldMineralX > silverMineral2X) {
//                                telemetry.addData("Gold Mineral Position", "Right");
//                            } else {
//                                telemetry.addData("Gold Mineral Position", "Center");
//                            }
//                        }
//                    }
//                    telemetry.update();
//                }
//            }
//        }
//
//        if (tfod != null) {
//            tfod.shutdown();
//        }
//    }
//}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumDrive.kt
package org.firstinspires.ftc.teamcode

import com.qualcomm.robotcore.eventloop.opmode.Autonomous
import com.qualcomm.robotcore.eventloop.opmode.Disabled
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.Servo
import com.qualcomm.robotcore.hardware.DcMotorSimple
import org.firstinspires.ftc.robotcore.external.ClassFactory
import org.firstinspires.ftc.robotcore.external.navigation.VuforiaLocalizer
import org.firstinspires.ftc.robotcore.external.tfod.TFObjectDetector

internal class MecanumDrive(hardwareMap: HardwareMap) {
    companion object {
        const val COUNTS_PER_ROTATION: Int = 657
    }
    private val frontLeft: DcMotor = hardwareMap.get(DcMotor::class.java, "leftFront")
    private val frontRight: DcMotor = hardwareMap.get(DcMotor::class.java, "rightFront")
    private val rearLeft: DcMotor = hardwareMap.get(DcMotor::class.java, "leftRear")
    private val rearRight: DcMotor = hardwareMap.get(DcMotor::class.java, "rightRear")
    private val arm: DcMotor = hardwareMap.get(DcMotor::class.java, "armRotate")
    private val lift: DcMotor = hardwareMap.get(DcMotor::class.java, "lift")
    private val dump: DcMotor = hardwareMap.get(DcMotor::class.java, "armExtend")
    private val collector: DcMotor = hardwareMap.get(DcMotor::class.java, "collectorIntake")
    // private val marker: Servo = hardwareMap.get(Servo::class.java, "marker")

    init {
        frontRight.direction = DcMotorSimple.Direction.REVERSE
        rearRight.direction = DcMotorSimple.Direction.REVERSE

        lift.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        arm.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE

        arm.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
    }

    fun drive(distance: Double, left: Double, right: Double) {
        frontLeft.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        rearLeft.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        frontRight.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        rearRight.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER

        frontLeft.mode = DcMotor.RunMode.RUN_TO_POSITION
        rearLeft.mode = DcMotor.RunMode.RUN_TO_POSITION
        frontRight.mode = DcMotor.RunMode.RUN_TO_POSITION
        rearLeft.mode = DcMotor.RunMode.RUN_TO_POSITION

        frontLeft.power = left
        rearLeft.power = left
        frontRight.power = right
        rearRight.power = right

        val leftDist = (if (left < 0.0) -distance else distance) * COUNTS_PER_ROTATION
        val rightDist = (if (right < 0.0) -distance else distance) * COUNTS_PER_ROTATION

        frontLeft.targetPosition = leftDist.toInt()
        rearLeft.targetPosition = leftDist.toInt()
        frontRight.targetPosition = rightDist.toInt()
        rearRight.targetPosition = rightDist.toInt()
    }

    fun driveMec(x: Double, y: Double, r: Double) {
        frontLeft.power = x + y + r
        frontRight.power = x - y - r
        rearLeft.power = x - y + r
        rearRight.power = x + y - r
    }


    fun setArmPower(pow: Double) {
        arm.power = pow
    }

    fun setDumpPower(pow: Double) {
        dump.power = pow
    }

    fun setLiftPower(pow: Double) {
        lift.power = pow
    }

    fun setCollectPower(pow: Double) {
        collector.power = pow
    }

    fun setMarkerPosition(position: Double) {
        // marker.position = position
    }
}

@Autonomous(name = "AutoCraterNative", group = "Autonomous")
@Disabled
internal class MecanumAutonomusHang : LinearOpMode() {
    companion object {
        private const val TFOD_MODEL_ASSET = "RoverRuckus.tflite"
        private const val LABEL_GOLD_MINERAL = "Gold Mineral"
        private const val LABEL_SILVER_MINERAL = "Silver Mineral"
        private const val VUFORIA_KEY = "AWalSLT/////AAABme4BfTt4y0Wnrn7kzOornkwjbHAMMZnWYYe0WOYGLbmQ914wZ8gozbWoSgsLHUQ+asiEx9VUOuDMkr4LsB5hB3iyn/JZM7NB1fG15dZSIXVDXevQ2iy+6zaVS8q0JZ2dpvugMWHEpsrmadtlA13znd7nTwhvVoc+2gGc2hNb5k6G4qe4l3jusYi3KGXTkzLNvjjYuQGP8xsBkGkXY3oHt6LNwdBy9EAcxRYmKEQUoGOOSNAL5vqB2ZHN2pKrlxXJs0BMbyhm1U4PTPYriGCb2rBmgiBvcJmfz9T/vVLCpL3T9mV1GByU38J1aMzj6QrQKWD6/ImZYMchpb2yG5Xq0vb1WyFGR9IKp6eaAJ0XQL+q"
    }

    private var lastDetection: String? = null
    private var tfod: TFObjectDetector? = null

    override fun runOpMode() {
        val driver = MecanumDrive(hardwareMap)

        run {
            val parameters = VuforiaLocalizer.Parameters()
            parameters.vuforiaLicenseKey = VUFORIA_KEY
            parameters.cameraDirection = VuforiaLocalizer.CameraDirection.BACK
            val vuforia = ClassFactory.getInstance().createVuforia(parameters)
            val tfodMonitorViewId = hardwareMap.appContext.resources.getIdentifier(
                    "tfodMonitorViewId", "id", hardwareMap.appContext.packageName)
            val tfodParameters = TFObjectDetector.Parameters(tfodMonitorViewId)
            tfodParameters.minimumConfidence = 0.30
            tfod = ClassFactory.getInstance().createTFObjectDetector(tfodParameters, vuforia)
            tfod!!.loadModelFromAsset(TFOD_MODEL_ASSET, LABEL_GOLD_MINERAL, LABEL_SILVER_MINERAL)
            tfod!!.activate()
        }

        val mineralPosition = runRecognition()
        tfod!!.deactivate()
        tfod!!.shutdown()

        waitForStart()

        // lower down
        driver.setLiftPower(-1.0)
        sleep(4300)
        driver.setLiftPower(0.0)

        // unlatch
        driver.driveMec(0.0, -0.5, 0.0)
        sleep(200)
        driver.driveMec(0.0, 0.0, 0.0)

        driver.setLiftPower(1.0)
        driver.setArmPower(1.0)

        // sample
        when (mineralPosition) {
            "left", null -> {
                driver.drive(0.3, 0.5, 0.5)
            }
            "right" -> {
                driver.drive(0.25, 0.5, 0.5)
            }
            "center" -> {
                driver.drive(0.65, 0.5, 0.5)
            }
        }

        driver.setLiftPower(0.0)
        driver.setArmPower(0.0)

        driver.drive(0.31, 0.5, 0.5)
        driver.drive(1.43, 0.5, -0.5)

//        val depotDrivePower = 0.7
//        when (mineralPosition) {
//            "left", null -> {
//                driver.drive(3.8, -depotDrivePower, -depotDrivePower)
//            }
//            "right" -> {
//                driver.drive(3.5, -depotDrivePower, -depotDrivePower)
//            }
//            "center" -> {
//                driver.drive(3.3, -depotDrivePower, -depotDrivePower)
//            }
//        }
//        sampleForwardAndBack(driver)

//        driver.drive(0.7, -0.5, 0.5)
//        driver.drive(1.8, 0.5, -0.5)
//
//        when (mineralPosition) {
//            "left", null -> {
//                driver.drive(3.8, -depotDrivePower, -depotDrivePower)
//            }
//            "right" -> {
//                driver.drive(3.0, -depotDrivePower, -depotDrivePower)
//            }
//            "center" -> {
//                driver.drive(3.2, -depotDrivePower, -depotDrivePower)
//            }
//        }
//
//        sleep(200)
//
//        driver.setMarkerPosition(-1.0)
//        sleep(2000)
//        driver.setMarkerPosition(0.5)
//
//        sleep(1000)
//
//        driver.drive(0.3, 0.5, -0.5)
//        driver.drive(4.75, 1.0, 1.0)
//
//        driver.setArmPower(0.5)
//        sleep(400)
//        driver.setArmPower(0.0)
    }

    private fun sampleForwardAndBack(driver: MecanumDrive) {
        driver.drive(0.69, 0.3, 0.3)
        driver.setCollectPower(1.0)
        sleep(1500)
        driver.setCollectPower(0.0)
        sleep(600)
    }

    private fun eagerRunRecognition(): String? {
        val updatedRecognitions = tfod?.updatedRecognitions
        if (updatedRecognitions != null) {
            telemetry.addData("# Object Detected", updatedRecognitions.size)
            var goldMineralX: Int? = null
            for (recognition in updatedRecognitions) {
                if (recognition.label == LABEL_GOLD_MINERAL) {
                    goldMineralX = recognition.right.toInt()
                }
            }
            if (goldMineralX == null) {
                return "right"
            }
            if (goldMineralX > 400) {
                return "center"
            }
            return "left"
        }
        return null
    }

    private fun runRecognition(): String? {
        val current = eagerRunRecognition()
        if (current != null) {
            lastDetection = current
        }
        telemetry.addData("detection", lastDetection)
        return lastDetection
    }
}